src/Orleans.Core.Abstractions/Runtime/RequestContext.cs

```diff
@@ -20,34 +20,49 @@ namespace Orleans.Runtime
     /// </remarks>
     public static class RequestContext
     {
-        /// <summary>
-        /// Gets or sets a value indicating whether <c>Trace.CorrelationManager.ActivityId</c> settings should be propagated into grain calls.
-        /// </summary>
-        public static bool PropagateActivityId { get; set; }
-
-        internal const string CALL_CHAIN_REQUEST_CONTEXT_HEADER = "#RC_CCH";
-        internal const string E2_E_TRACING_ACTIVITY_ID_HEADER = "#RC_AI";
+        internal const string CALL_CHAIN_REENTRANCY_HEADER = "#CCR";
         internal const string PING_APPLICATION_HEADER = "Ping";
 
         internal static readonly AsyncLocal<ContextProperties> CallContextData = new AsyncLocal<ContextProperties>();
 
-        /// <summary>Gets or sets an activity ID that can be used for correlation.</summary>
-        public static Guid ActivityId
+        public static Guid ReentrancyId
         {
-            get { return (Guid)(Get(E2_E_TRACING_ACTIVITY_ID_HEADER) ?? Guid.Empty); }
+            get => Get(CALL_CHAIN_REENTRANCY_HEADER) is Guid guid ? guid : Guid.Empty;
             set
             {
                 if (value == Guid.Empty)
                 {
-                    Remove(E2_E_TRACING_ACTIVITY_ID_HEADER);
+                    Remove(CALL_CHAIN_REENTRANCY_HEADER);
                 }
                 else
-                { 
-                    Set(E2_E_TRACING_ACTIVITY_ID_HEADER, value);
+                {
+                    Set(CALL_CHAIN_REENTRANCY_HEADER, value);
                 }
             }
         }
 
+        public static ConfiguredCallChain AllowCallChainReentrancy()
+        {
+            var originalCallChainId = ReentrancyId;
+            if (originalCallChainId == Guid.Empty)
+            {
+                ReentrancyId = Guid.NewGuid();
+            }
+
+            return new ConfiguredCallChain(originalCallChainId);
+        }
+
+        public static ConfiguredCallChain SuppressCallChainReentrancy()
```

Comment:
Ik now that we have direct access to the `ReentrancyId`, but if we do return a `ConfiguredCallChain`, maybe we should have a `RestoreConfiguredCallChain(ConfiguredCallChain)` to use it if necessary

Reply:
It's supposed to be used as a using variable so that `Dispose()` will be called.
```
using (RequestContext.SuppressCallChainReentrancy())
{
...
}
```
or
``` C#
{
  using var _ = RequestContext.SuppressCallChainReentrancy();
  ...
}
```